### PR TITLE
KOGITO-6864: [SWF Diagram Editor] Support for YAML workflow declaration

### PR DESCRIPTION
**Pull request by treblereel**
jira: https://issues.redhat.com/browse/KOGITO-6864

### CODE REVIEW

**Comment by caponetto** [comment]
@treblereel can you please add a gif to show off this new feature?

**Comment by treblereel** [comment]
> @treblereel can you please add a gif to show off this new feature?

Screen Recording 2023-02-07 at 10 26 27 PM

**Comment by romartin** [comment]
Hey @treblereel 

Starting the code review, sorry for the delay.

BTW can you please rebase and get a green build again, so we can use the generated artifacts for testing? 

Thanks!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/src/main/webapp/test.html [hunk 1]
```diff
@@ -112,499 +156,357 @@
       const examples = [
         {
           id: "injectExample",
-          version: "1.0",
```
Comment (romartin) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/src/main/webapp/test.html:115]: I understand you replaced all `JSON` workflow declarations by `YAML` ones, but would it make sense to preserve both? This way, while dev. phase, we can quickly test any format. Maybe as you suggested, 2 dropdowns for examples in JSON & YAML? WDYT?

File: packages/serverless-workflow-diagram-editor/uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/EditorActivity.java [hunk 2]
```diff
@@ -26,12 +26,16 @@ public interface EditorActivity extends Activity {
 
     Promise<Void> setContent(String path, String value);
 
+    Promise<Void> setContentYAML(String path, String value);
```
Comment (romartin) [inline packages/serverless-workflow-diagram-editor/uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/EditorActivity.java]: As far as I know our tooling's multiplying architecture does not support these new methods for GWT Editors (`set/getContentYAML`). So I understand that actually there are no calls to these methods neither.

Instead, the SWF editor (stunner based) must handle the `JSON`/`YAML` format "internally". So from the tooling perspective, we have to handle formats for the `setContent()` and `getContent()` methods in our editor class. This way:

- once setting a JSON content, parsers will use JSON format and also the getContent result must be JSON
- once setting a YAML content, parsers will use JSON format and also the getContent result must be JSON
- So I understand this `EditorActivity` class will remain with no changes finally

**Review by romartin — CHANGES_REQUESTED** [review]
Hey @treblereel 

Thanks for this work! Looks great! :-D

I did code review for all changes files except the ones from the `third_party` folder. In general changes are looking good to me, just added a few review comments.

Anyway, still asking for **3 key changes**:

1. The GWT Editor class must **handle content for both `JSON`/`YAML` formats**. This way, we have to remain exposing only `setContent`/`getContent` methods. See my review comment below for more details.

2. Once testing the VSCode extension artifact generated from the build, and creating a new `.sw.yaml` file. **it sill shows default mermaid editor, instead of our SWF one** (stunner based). This is expected because we were not supporting YAML for displaying the workflow. But at this point, I think this PR must change this, otherwise QE will be not also able to test the artifacts. Can you please ask @paulovmr  about this? It is implemented in some package of the typescript modules in the kie-tools.

3. At the end, would be also great to have **automated integration tests** for both JSON / YAML. Actually there exist several tests for JSON files, we can run the same test cases but also considering YAML.

Rather than this, from that point, we will be able to start QE review and provide more feedback. Anyway this is looking great, let's see if we can move this forward quicly. Let me know if something doesn't make sense for you guys!

Thanks!

**Comment by romartin** [comment]
Hey @paulovmr , can you point to some files to @treblereel about how to remove the use of mermaid editor, when using YAML, please? See my comment above, point 2 ^^

Thanks in advance!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/src/main/webapp/test.html (hunk 1 above)
Comment (treblereel) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/src/main/webapp/test.html:115]: done

File: packages/serverless-workflow-diagram-editor/uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/EditorActivity.java (hunk 2 above)
Comment (treblereel) [reply packages/serverless-workflow-diagram-editor/uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/EditorActivity.java]: done

**Comment by treblereel** [comment]
@paulovmr, we need your wisdom, could you help us with VSCode extension?

**Comment by treblereel** [comment]
> The (build) kie-tools-stunner-editors check has **failed**. Please check [the logs](https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/kie-tools.tests.kie-tools-stunner-editors/1505/display/redirect).

failed to initialize the tests

**Comment by caponetto** [comment]
Hi @romartin @treblereel 
We have a feature flag to enable/disable stunner for JSON files.
IMO now we can use this feature flag to enable/disable stunner for both JSON and YAML files. 
WDYT? I can send a PR to this branch.

**Comment by treblereel** [comment]
> Hi @romartin @treblereel We have a feature flag to enable/disable stunner for JSON files. IMO now we can use this feature flag to enable/disable stunner for both JSON and YAML files. WDYT? I can send a PR to this branch.

sounds good for me

**Comment by caponetto** [comment]
@treblereel there you go
https://github.com/treblereel/kie-tools/pull/1

**Comment by LuboTerifaj** [comment]
Hello @treblereel , @romartin , please let me know once the PR is ready for qe review. Thank you!

**Comment by romartin** [comment]
Hey @treblereel 

How's progress here? I see new commits but not sure if this is ready for review. Please let us know when ready!!

Thanks!

**Comment by treblereel** [comment]
> Hey @treblereel 
> 
> How's progress here? I see new commits but not sure if this is ready for review. Please let us know when ready!!
> 
> Thanks!

The only issue left is default mermaid editor, but looks like @paulovmr is on PTO.

**Comment by paulovmr** [comment]
Hi @treblereel , I believe that the changes @caponetto did in his PR to your branch should fix that issue. So, if the flag is enabled, the stunner editor should open for YAML workflow files. I sent another PR to your branch to change just the description of the flag as well: https://github.com/treblereel/kie-tools/pull/2.

**Comment by romartin** [comment]
great, thanks @caponetto and @paulovmr ! 

Please @treblereel can you please check with both PRs you can open YAML workflows with stunner?

**Comment by treblereel** [comment]
> great, thanks @caponetto and @paulovmr !
> 
> Please @treblereel can you please check with both PRs you can open YAML workflows with stunner?

yeap, it works

**Comment by romartin** [comment]
Hey @treblereel 

About the **failed CI build for Ubuntu**.

I downloaded the full log and I see **a failed execution for some integration test**. By looking at other PRs, this test works, so can you please verify your PR is passing tests properly? 

Thanks!

**Error from log:**
`2023-03-20T21:59:29.8462814Z Writing code settings to /home/runner/work/kie-tools/kie-tools/packages/vscode-extension-serverless-workflow-editor/test-resources/settings/User/settings.json
2023-03-20T21:59:29.8467695Z Launching browser...
2023-03-20T21:59:36.5730396Z Browser ready in 6728 ms
2023-03-20T21:59:36.5738442Z Launching tests...
2023-03-20T21:59:36.5742708Z 
2023-03-20T21:59:36.5746400Z   Serverless workflow editor - smoke integration tests
2023-03-20T21:59:52.5197198Z       ✔ Opens greetings.sw.json and loads two editor groups: 10423ms
2023-03-20T21:59:55.4682413Z 
2023-03-20T21:59:55.4689919Z   Suite duration: 18.894 s, Tests: 1
2023-03-20T22:00:00.5878922Z 
2023-03-20T22:00:00.5885006Z   Serverless workflow editor - autocompletion tests JSON files
2023-03-20T22:00:36.9781637Z       ✔ Completes serverless workflow with function and state autocompletion: 36112ms
2023-03-20T22:00:42.7909913Z       1) "before each" hook for "Completes serverless workflow from an empty file"
2023-03-20T22:00:44.4507723Z 
2023-03-20T22:00:44.4556928Z   Suite duration: 43.864 s, Tests: 2
2023-03-20T22:00:44.4566303Z 
2023-03-20T22:00:44.4569799Z   Serverless workflow editor - Basic operations tests
2023-03-20T22:01:23.3515108Z       1) "after each" hook for "Opens, edits and saves the *.sw.json file"
2023-03-20T22:01:43.2802758Z Shutting down the browser
2023-03-20T22:01:53.7076648Z Uninstalling redhat.vscode-extension-serverless-workflow-editor...
2023-03-20T22:01:55.0662804Z Extension 'redhat.vscode-extension-serverless-workflow-editor' was successfully uninstalled!
2023-03-20T22:01:55.2476185Z       2) "after all" hook in "{root}"
2023-03-20T22:01:55.2493843Z 
2023-03-20T22:01:55.2501881Z   Suite duration: 70.793 s, Tests: 2
2023-03-20T22:01:55.2546208Z 
2023-03-20T22:01:55.2557538Z   2 passing (2m)
2023-03-20T22:01:55.2561058Z   3 failing
2023-03-20T22:01:55.2563025Z 
2023-03-20T22:01:55.2575047Z   1) Serverless workflow editor - autocompletion tests
2023-03-20T22:01:55.2575999Z        "before each" hook for "Completes serverless workflow from an empty file":
2023-03-20T22:01:55.2580495Z      Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/runner/work/kie-tools/kie-tools/packages/vscode-extension-serverless-workflow-editor/out/serverless-workflow-editor-extension-autocompletion.test.js)
2023-03-20T22:01:55.2586608Z       at listOnTimeout (node:internal/timers:569:17)
2023-03-20T22:01:55.2587287Z       at process.processTimers (node:internal/timers:512:7)
2023-03-20T22:01:55.2590846Z 
2023-03-20T22:01:55.2592091Z   2) "after each" hook for "Opens, edits and saves the *.sw.json file":
2023-03-20T22:01:55.2598278Z      Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
2023-03-20T22:01:55.2612594Z       at listOnTimeout (node:internal/timers:569:17)
2023-03-20T22:01:55.2617539Z       at process.processTimers (node:internal/timers:512:7)
2023-03-20T22:01:55.2621312Z 
2023-03-20T22:01:55.2621521Z   3) "after all" hook in "{root}":
2023-03-20T22:01:55.2645201Z      Error: Timeout of 30000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
2023-03-20T22:01:55.2663498Z       at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
2023-03-20T22:01:55.2670092Z 
2023-03-20T22:01:55.2671357Z 
2023-03-20T22:01:55.2671371Z 
2023-03-20T22:01:56.0692644Z [run-script-if] Error executing 'extest setup-and-run --yarn -c max -u -e ./test-resources -o ./it-tests/settings.json out/*.test.js'.`

**Comment by treblereel** [comment]
retest this please

**Comment by romartin** [comment]
Hey @treblereel 

I tested locally:
1.- checkout your branch `yaml_marshaller`
2.- rebase from main
3.- cherry-pick this fix from Paulo -> https://github.com/kiegroup/kie-tools/pull/1536/commits/484db5fd1c84ff26fef7595ead9b95548447613c#diff-4dffe89447a3479f6fe480860ff1f7f14a101bafc177085963c682c8e53c5524
3.- fix some rebase conflicts on DiagramEditor.java / DiagramEditorTest.java
4.- Then completely DISABLED `SWEditorYamlSeleniumIT` test (they fail, not sure why)
5.- Run build:prod with tests
and it results now on some integration tests failed on YAML parsing
see https://gist.github.com/romartin/59c86ed986d8ab6256edb558ac9dc4d1

This way I think you should work on fixing thosre 2 errors now:
1.- **SWEditorYamlSeleniumIT** deson't work
2.-**Serverless workflow editor - autocompletion tests** doesn't work for YAML (integration test in vscode-extension)


Does it makes sense?

thanks!

**Comment by treblereel** [comment]
@romartin 

On my local fedora tests are failing on:
```
 1) Serverless workflow editor - autocompletion tests
       "after each" hook for "Completes serverless workflow from an empty file":
     StaleElementReferenceError: stale element reference: element is not attached to the page document
  (Session info: chrome=102.0.5005.196)
      at Object.throwDecodedError (/home/treblereel/workspace/redhat/kie-tools/node_modules/.pnpm/selenium-webdriver@4.8.1/node_modules/selenium-webdriver/lib/error.js:524:15)
      at parseHttpResponse (/home/treblereel/workspace/redhat/kie-tools/node_modules/.pnpm/selenium-webdriver@4.8.1/node_modules/selenium-webdriver/lib/http.js:587:13)
      at Executor.execute (/home/treblereel/workspace/redhat/kie-tools/node_modules/.pnpm/selenium-webdriver@4.8.1/node_modules/selenium-webdriver/lib/http.js:515:28)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Driver.execute (/home/treblereel/workspace/redhat/kie-tools/node_modules/.pnpm/selenium-webdriver@4.8.1/node_modules/selenium-webdriver/lib/webdriver.js:745:17)

  2) Serverless workflow editor - functions tests
       "after each" hook for "Checks functions are loaded from specs and routes directories into YAML serverless workflow file":
     StaleElementReferenceError: stale element reference: element is not attached to the page document
  (Session info: chrome=102.0.5005.196)
      at Object.throwDecodedError (/home/treblereel/workspace/redhat/kie-tools/node_modules/.pnpm/selenium-webdriver@4.8.1/node_modules/selenium-webdriver/lib/error.js:524:15)
      at parseHttpResponse (/home/treblereel/workspace/redhat/kie-tools/node_modules/.pnpm/selenium-webdriver@4.8.1/node_modules/selenium-webdriver/lib/http.js:587:13)
      at Executor.execute (/home/treblereel/workspace/redhat/kie-tools/node_modules/.pnpm/selenium-webdriver@4.8.1/node_modules/selenium-webdriver/lib/http.js:515:28)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Driver.execute (/home/treblereel/workspace/redhat/kie-tools/node_modules/.pnpm/selenium-webdriver@4.8.1/node_modules/selenium-webdriver/lib/webdriver.js:745:17)
```

I am not really familiar with those tests, looks like some dom element was removed or something, so I need assistance to fix it ....

**Comment by romartin** [comment]
@handreyrc @paulovmr you have any clue about what "stale element" means in the test failure? Any ideas? See ^^^

**Comment by romartin** [comment]
anyway @treblereel , until https://github.com/kiegroup/kie-tools/pull/1536 doesn't get merged, this PR should not properly pass the tests, isn't it? Couldn't it be related?

**Comment by treblereel** [comment]
@romartin not sure, fwiw these two errors aren't connected to the pr you mentioned

**Comment by treblereel** [comment]
possible fix https://github.com/treblereel/kie-tools/pull/3

**Comment by treblereel** [comment]
retest this please

**Comment by romartin** [comment]
Hey @treblereel @LuboTerifaj 

First of all, thanks @treblereel for fixing all code review items I previously added. I have been also testing the generated VSIX, and it works good! For now I found no issues. what I am focused on testing is:
1.- Some YAML examples (.sw.yaml files)
2.- Checking the information shown in each state's detail (icons and tooltips) is correct

@LuboTerifaj As discussed, although there exista the unrelated integration test issue, this work is **TOP PRIORITY**, so would appreciate proper prioritization please. 

PS: Notice the goal is just testing the vsocde extension, so just opening json or yaml files, and it IS NOT about saving (serializing) files. 
PS: Also notice @treblereel added integration tests for YAML workflows :)

Thanks guys!! Great work!!

**Review by romartin — APPROVED** [review]

**Comment by romartin** [comment]
Hey @treblereel 

Sorry for the inconveniences but we just merged a quite some big PR, so you'll have **to rebase** (hopefully for last time!).

Anyway @LuboTerifaj this is ready for QE testing. Artifacts are available on the checks tab, as usually. So would appreciate your early reivew and feedback here.

Thanks guys!

**Comment by treblereel** [comment]
God no

**Comment by romartin** [comment]
thanks for the rebase @treblereel !! 

Just done some smoketesting for generated _vsix_ and all good, ready for review @LuboTerifaj !

**Review by LuboTerifaj — CHANGES_REQUESTED** [review]
Hello @treblereel , @romartin,

I found the following issues:

"Editor content has not been set" is shown instead of expected Diagram. See the following examples from specification:

- [deliver-order-sub-workflow.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/online-food-ordering/deliver-order-sub-workflow.sw.yaml)
- [main-workflow-definition.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/online-food-ordering/main-workflow-definition.sw.yaml)
- [place-order-sub-workflow.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/online-food-ordering/place-order-sub-workflow.sw.yaml)
- [accumulate-room-readings.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/accumulate-room-readings.sw.yaml)
- [applicant-request-decision.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/applicant-request-decision.sw.yaml)
- [check-inbox-periodically.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/check-inbox-periodically.sw.yaml)
- [event-based-greeting.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/event-based-greeting.sw.yaml)
- [event-based-transitions.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/event-based-transitions.sw.yaml)
- [finalize-college-application.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/finalize-college-application.sw.yaml)
- [greeting.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/greeting.sw.yaml)
- [handle-car-auction-bids.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/handle-car-auction-bids.sw.yaml)
- [monitor-patient-vital-signs.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/monitor-patient-vital-signs.sw.yaml)
- [perform-customer-credit-check.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/perform-customer-credit-check.sw.yaml)
- [purchase-order-deadline.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/purchase-order-deadline.sw.yaml)
- [reusing-function-and-event-definitions.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/reusing-function-and-event-definitions.sw.yaml)
- [send-cloudEvent-on-workflow-completion.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/send-cloudEvent-on-workflow-completion.sw.yaml)
- [solving-math-problems.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/solving-math-problems.sw.yaml)

There is missing end node in diagram:
- [parallel-execution.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/parallel-execution.sw.yaml)

There is a parser message in Text Editor shown at the 24th line in the following workflow:
- [new-patient-onboarding.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/new-patient-onboarding.sw.yaml)
  - `Incorrect type. Expected "action".yaml-schema: Event State`
  
  
 Note, I may find also more issues during testing. I will let you know if I find anything next week.

Thank you!

**Comment by treblereel** [comment]
@LuboTerifaj thanks for your findings. Do you think [deliver-order-sub-workflow.sw.yaml](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/online-food-ordering/deliver-order-sub-workflow.sw.yaml) is a valid yaml?

**Comment by LuboTerifaj** [comment]
Hello @treblereel , yes it is a valid yaml file.

**Comment by romartin** [comment]
ok thanks @LuboTerifaj ! We are working on those fixes :+1:

**Comment by romartin** [comment]
@treblereel it looks the test `UploadFiles.cy.ts` is failing, under the `serverless-logic-webtools` package

**Review by LuboTerifaj — APPROVED** [review]
Hello @treblereel ,

I didn't find any more significant issues.

The one that should be fixed after this PR:

- In [ examples/resources/online-food-ordering/main-workflow-definition.sw.yaml ](https://github.com/LuboTerifaj/specification/tree/examples-in-separate-files/examples/resources/online-food-ordering/main-workflow-definition.sw.yaml ), the following element is not shown in the diagram properly: `output: '${ . | {"orderid": .id, "orderstatus": .status} | .orderstatus += ["Order Completed"] }'`

There is related issue in the sw-editor-kogito-app, but again, not for this PR:

- There is json suffix instead of yaml in downloaded files using "the diagram as loaded" feature in the webapp.

Other than that, it looks good to me.

Thank you for the fixes and the great work on this!

**Comment by LuboTerifaj** [comment]
I have reported those issues here:

- [KOGITO-9053](https://issues.redhat.com/browse/KOGITO-9053) _[Diagram Viewer] stateDataFilter doesn't show details in diagram properly_
- [KOGITO-9054](https://issues.redhat.com/browse/KOGITO-9054) _Wrong suffixes in downloaded files from sw-editor-kogito-app_

**Pull request merged**